### PR TITLE
This makes the menu behave more like menus of other gui frameworks

### PR DESCRIPTION
**Pull request by DanielKeogh**
An open menu will now no longer close during the first click when the mouse-button is released unless it is released over a button with a command.
Clicking on an open menu will close it, as will clicking outside of the menu.

To support this, it was necessary to track which menu was open.
The active menu is now tracked on the standard-application-frame in a slot named 'active-menu'.
I figured that this was an ok design choice, as the standard-application-frame already has two menu-specific slots on it.

### CODE REVIEW

File: Core/clim-core/menu.lisp [hunk 1]
```diff
@@ -170,7 +175,10 @@ account, and create a list of menu buttons."
 	    (adopt-frame manager submenu-frame)
 	    (enable-frame submenu-frame)
 	    (with-sheet-medium (medium raised)
-	      (medium-force-output medium))))))))
+	      (medium-force-output medium))))))
+    (when frame
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: functions earlier would signal a condition if frame is  nil, so there is no need for this test

File: Core/clim-core/menu.lisp [hunk 2]
```diff
@@ -50,6 +50,11 @@
       (dispatch-repaint button (sheet-region button))
       (stream-force-output button))))
 
+(defgeneric toggle-arm-branch (button))
+
+(defmethod toggle-arm-branch ((button menu-button-pane))
+  (arm-menu button))
+
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: please rename arm-menu function to toggle-arm-branch and merge their behaviors. also write a comment above defgeneric form which describe how we achieve desired behavior.

File: Core/clim-core/menu.lisp [hunk 3]
```diff
@@ -179,7 +187,10 @@ account, and create a list of menu buttons."
       (disown-frame frame-manager submenu-frame)
       (disarm-gadget sub-menu)
       (dispatch-repaint sub-menu +everywhere+)
-      (setf submenu-frame nil) )))
+      (setf submenu-frame nil)))
+  (with-application-frame (frame)
+    (when frame
+      (setf (slot-value frame 'active-menu) nil))))
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: make active-menu an accessor on frame (here you use setf on slot-value and earlier you use with-slots - this is not consistent).

File: Core/clim-core/menu.lisp [hunk 4]
```diff
@@ -192,8 +203,19 @@ account, and create a list of menu buttons."
...
-(defmethod handle-event ((pane menu-button-submenu-pane) (event pointer-button-release-event))
-  (destroy-substructure (menu-root pane)))
+(defmethod toggle-arm-branch ((button menu-button-submenu-pane))
+  (if (slot-value button 'armed)
+      (destroy-substructure button)
+      (arm-branch button)))
+
+(defmethod handle-event :before (pane (event pointer-button-release-event))
+  (when (not (or (typep event 'window-repaint-event)
+		  (typep event 'pointer-motion-event)))
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: s/when not/unless/

File: Core/clim-core/menu.lisp [hunk 5]
```diff
@@ -192,8 +203,19 @@ account, and create a list of menu buttons."
...
+    (with-application-frame (frame)
+      (when (and (not (typep pane 'menu-button)) frame)
+	(with-slots (active-menu) frame
+	  (when active-menu
+	    (destroy-substructure (menu-root active-menu))))))))
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]:
```
(unless (typep pane 'menu-button)
  (with-application-frame (frame)
    (when (and frame (active-menu frame))
      (destroy-sub …)
```

File: Core/clim-core/frames.lisp [hunk 6]
```diff
@@ -113,6 +113,7 @@ input focus. This is a McCLIM extension."))
 	     :initform nil)
    (menu-bar-pane :initform nil
                   :accessor frame-menu-bar-pane)
+   (active-menu :initform nil)
```
Comment (dkochmanski) [inline Core/clim-core/frames.lisp]: make it an accessor %frame-active-menu

**Comment by DanielKeogh** [comment]
Will do these things. Thanks for review.

After this I fill fix the weird double-open that occurs when you mouse-down a menu-item that has no sub-menu, and then drag the mouse over a menu-item that does, then release.
(e.g. in menu demo, press down on right menu, then release over left.)

Without this pullrq you are left with an open menu that can only be closed by clicking on it directly.

File: Core/clim-core/menu.lisp (hunk 2 above)
Comment (DanielKeogh) [reply Core/clim-core/menu.lisp]: Can't merge arm-menu and toggle-arm-branch.
arm-menu when menu is already armed. toggle-arm-branch is solely for pointer-button-press-event.

I can remove toggle-arm-branch by overriding handle-event for menu-button-submenu-pane.

File: Core/clim-core/frames.lisp (hunk 6 above)
Comment (DanielKeogh) [reply Core/clim-core/frames.lisp]: Done.

File: Core/clim-core/menu.lisp (hunk 4 above)
Comment (DanielKeogh) [reply Core/clim-core/menu.lisp]: Done

File: Core/clim-core/menu.lisp (hunk 1 above)
Comment (DanielKeogh) [reply Core/clim-core/menu.lisp]: Okies

File: Core/clim-core/menu.lisp (hunk 3 above)
Comment (DanielKeogh) [reply Core/clim-core/menu.lisp]: In the other place I accessed the slot twice. Here only once. I try to always use with-accessors/with-slots in any place where there are multiple uses, and avoid it when there is only one.
I changed this to use the accessor in this case.

File: Core/clim-core/menu.lisp (hunk 5 above)
Comment (DanielKeogh) [reply Core/clim-core/menu.lisp]: Changed, thanks.

**Comment by DanielKeogh** [comment]
I left two feedbackos unresolved.
I'm not too happy with the existence of:
`(defmethod handle-event :before (pane (event pointer-button-release-event)) ...)`
Given that it adds overhead to every pointer-button-release-event, but like I said in the irc channel, there aren't any hooks for adding temporary event handlers at this stage.

File: Core/clim-core/menu.lisp [hunk 7]
```diff
@@ -192,8 +196,17 @@ account, and create a list of menu buttons."
...
     (arm-menu sub-menu)))
 
-(defmethod handle-event ((pane menu-button-submenu-pane) (event pointer-button-release-event))
-  (destroy-substructure (menu-root pane)))
+(defmethod handle-event ((pane menu-button-submenu-pane) (event pointer-button-press-event))
+  (if (slot-value pane 'armed)
+      (destroy-substructure pane)
+      (arm-branch pane)))
+
+(defmethod handle-event :before (pane (event pointer-button-release-event))
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: why isn't this specialized on pane as `menu-button-pane`?

Comment (dkochmanski) [reply Core/clim-core/menu.lisp]: then handle-event would trigger only for panes of this type and there would be no unnecessary overhead

Comment (dkochmanski) [reply Core/clim-core/menu.lisp]: ah, nevermind.

**Comment by dkochmanski** [comment]
I'm wondering if it wouldn't be a good idea to make menu be responsive to keyboard events (as in `port-keyboard-input-focus`), then we could use our (currently malfunctioning) protocol `note-input-focus-changed` and specialize for menu, so when it loses focus then it is closed. As a bonus we could implement navigating menu with arrows. What do you think?

nb: I've tested your branch and it works as advertised.  I have too much on my plate to work on what I've suggested above, so it is up to you: should we merge the branch or work on it further?

**Comment by DanielKeogh** [comment]
Clearly, menu navigation with arrows is a good idea.

Let me think about it and come back. I will need to think about is whether or not focus-change is the right abstraction for this to be hooked on.

I am expecting that the "focus" abstraction is not clearly defined as, "focus changes when something is clicked on", and some of the time focus changed events could be raised by a mouse-over or other shortcuts. (It would be extremely annoying if the menu was closed because of this)

If this is the case, then we would want something like:
`(defgeneric can-yield-focus (event-causing-focus-change item-with-focus item-to-be-focused))`
This would let us refuse to yield focus under certain circumstances. Notably, this would be useful for implementing modal dialog (pop-up that stops you from clicking the window behind it).

I will look at how this protocol works on the weekend and decide.

**Comment by dkochmanski** [comment]
I don't think we need to introduce new protocols. if menu loses keyboard focus then it should close. currently keyboard focus only changes upon click.

nb: I'm anxious about introducing more protocol complexity onto what we have now. One of my hazy plans for the future is to build a chinese walls between basic backend abstraction, windowing abstraction and application abstraction (ports/grafts vs sheets/streams vs frames/panes), so one is built on top of another instead of having a monolith.

**Comment by DanielKeogh** [comment]
The menu changes are right, I think.

The change to focus is very aggressive. I also haven't looked at the document on how focus is supposed to work yet.

**Comment by dkochmanski** [comment]
I'm closing this for now, if the pull request reemerge it should be done after we sort out #851.

**Pull request closed without merging**